summoner-cli/src/Summoner/Template/GitHub.hs

```diff
@@ -121,49 +121,121 @@ gitHubFiles Settings{..} = concat
         , "    branches: [master]"
         , ""
         , "jobs:"
-        , "  build:"
-        , "    name: ghc ${{ matrix.ghc }}"
-        , "    runs-on: ubuntu-16.04"
+        ]
+        <> memptyIfFalse settingsCabal ghActionsCabal
+        <> memptyIfFalse settingsStack ghActionsStack
+
+    ghcActionsCheckoutVersion = "@v2.3.3"
+    ghcActionsSetupHaskellVersion = "@v1.1.3"
+    ghcActionsCacheVersion = "@v2.1.2"
+
+    ghActionsCabal :: [Text]
+    ghActionsCabal =
+        [ "  cabal:"
+        , "    name: ${{ matrix.os }} / ghc ${{ matrix.ghc }}"
+        , "    runs-on: ${{ matrix.os }}"
         , "    strategy:"
         , "      matrix:"
+        , "        os: [ubuntu-latest, macOS-latest, windows-latest]"
         , "        cabal: [" <> quote defaultCabal <> "]"
         , "        ghc:"
         ]
         <> map (indent 10 <>) ghActionsVersions
+        <> ghActionsExcludes
         <>
         [ ""
         , "    steps:"
-        , "    - uses: actions/checkout@v2"
+        , "    - uses: actions/checkout" <> ghcActionsCheckoutVersion
         , "      if: github.event.action == 'opened' || github.event.action == 'synchronize' || github.event.ref == 'refs/heads/master'"
         , ""
-        , "    - uses: actions/setup-haskell@v1"
+        , "    - uses: actions/setup-haskell" <> ghcActionsSetupHaskellVersion
+        , "      id: setup-haskell-cabal"
         , "      name: Setup Haskell"
         , "      with:"
         , "        ghc-version: ${{ matrix.ghc }}"
         , "        cabal-version: ${{ matrix.cabal }}"
         , ""
-        , "    - uses: actions/cache@v1"
+        , "    - name: Freeze"
+        , "      run: |"
+        , "        cabal v2-freeze"
+        , ""
+        , "    - uses: actions/cache" <> ghcActionsCacheVersion
         , "      name: Cache ~/.cabal/store"
         , "      with:"
-        , "        path: ~/.cabal/store"
-        , "        key: ${{ runner.os }}-${{ matrix.ghc }}-cabal"
+        , "        path: ${{ steps.setup-haskell-cabal.outputs.cabal-store }}"
+        , "        key: ${{ runner.os }}-${{ matrix.ghc }}-${{ hashFiles('cabal.project.freeze') }}"
         , ""
         , "    - name: Build"
         , "      run: |"
-        , "        " <> cabalUpdate
+        , "        " <> cabalConfigure
         , "        " <> cabalBuild
         , ""
         , "    - name: Test"
         , "      run: |"
         , "        " <> cabalTest
+        , ""
```

Comment:
I think this is redundant 🙂

Reply:
I see that it is needed when both `cabal` and `stack` are used. But I think this case could be handled separately, and we may add an empty line only when necessary 🙂 

E.g.
```haskell
    <> memptyIfFalse (settingsCabal && settingsStack) [""]
```

Reply:
I will fix that

Reply:
Done